db/versions/0070.yml

```diff
@@ -0,0 +1,248 @@
+version: 70
+description: add `get_worker_pool_with_capacity_and_counts_by_state`, `get_worker_pools_with_capacity_and_counts_by_state`, and `update_worker_pool_with_capacity_and_counts_by_state` functions to get worker counts and capacity by state for worker pools
+methods:
+  get_worker_pool_with_capacity:
+    deprecated: true
+  get_worker_pools_with_capacity:
+    deprecated: true
+  update_worker_pool_with_capacity:
+    deprecated: true
+  get_worker_pool_with_capacity_and_counts_by_state:
+    description: |-
+      Get an existing worker pool.  The returned table will have one or (if no such worker pool is defined) zero rows.
+    mode: read
+    serviceName: worker_manager
+    args: worker_pool_id_in text
+    returns: table(worker_pool_id text, provider_id text, previous_provider_ids jsonb, description text, config jsonb, created timestamptz, last_modified timestamptz, owner text, email_on_error boolean, provider_data jsonb, current_capacity integer, requested_count integer, running_count integer, stopping_count integer, stopped_count integer, requested_capacity integer, running_capacity integer, stopping_capacity integer, stopped_capacity integer)
+    body: |-
+      begin
+        return query
+        select
+          worker_pools.worker_pool_id,
+          worker_pools.provider_id,
+          worker_pools.previous_provider_ids,
+          worker_pools.description,
+          worker_pools.config,
+          worker_pools.created,
+          worker_pools.last_modified,
+          worker_pools.owner,
+          worker_pools.email_on_error,
+          worker_pools.provider_data,
+          coalesce((
+            select sum(workers.capacity) from workers where
```

Comment:
Ok, so in this query, according to `Explain` there will be 1 main query + 9 sub-queries (although very fast ones, because they will be filtering by primary key)
<img width="445" alt="image" src="https://user-images.githubusercontent.com/83861/161821735-8452c6a7-bfef-4507-8e55-86ca6bd17714.png">


This can be rewritten to just have one join and calculation inside using `CASE WHEN`

```sql
        select
          worker_pools.worker_pool_id,
          worker_pools.provider_id,
          worker_pools.previous_provider_ids,
          worker_pools.description,
          worker_pools.config,
          worker_pools.created,
          worker_pools.last_modified,
          worker_pools.owner,
          worker_pools.email_on_error,
          worker_pools.provider_data,
          coalesce(sum(CASE WHEN workers.state != 'stopped' THEN workers.capacity ELSE 0 END))::integer as non_stopped_capacity,          
          coalesce(sum(CASE WHEN workers.state = 'running' THEN workers.capacity ELSE 0 END))::integer as running_capacity,
          coalesce(sum(CASE WHEN workers.state = 'requested' THEN workers.capacity ELSE 0 END))::integer as requested_capacity,
          coalesce(sum(CASE WHEN workers.state = 'stopping' THEN workers.capacity ELSE 0 END))::integer as stopping_capacity,
          coalesce(sum(CASE WHEN workers.state = 'stopped' THEN workers.capacity ELSE 0 END))::integer as requested_capacity,
          coalesce(count(CASE WHEN workers.state = 'requested' THEN 1 ELSE 0 END))::integer as requested_workers,
          coalesce(count(CASE WHEN workers.state = 'running' THEN 1 ELSE 0 END))::integer as running_workers,
          coalesce(count(CASE WHEN workers.state = 'stopping' THEN 1 ELSE 0 END))::integer as stopping_workers,
          coalesce(count(CASE WHEN workers.state = 'stopped' THEN 1 ELSE 0 END))::integer as stopped_workers
        from worker_pools
        left join workers ON workers.worker_pool_id = worker_pools.worker_pool_id
        group by worker_pools.worker_pool_id
        order by worker_pools.worker_pool_id
```

This results in two sequential scans
<img width="402" alt="image" src="https://user-images.githubusercontent.com/83861/161821821-5d00a4b5-d5b4-49dc-82ce-a07f88c2194e.png">

And it feels like doing second one is ~30-50% faster on community tc database (probably because query needs to scan all rows anyways)

Reply:
So it looks like single join is performing better and I think the other query should also be rewritten :+1:

Reply:
Updated with joins! https://github.com/taskcluster/taskcluster/pull/5340/files#diff-44414ecd761296d11428d72a4701ed1db030571c35858b90e4ac65f48d1ff7a2

Thanks for the help @lotas 👍🏻

Reply:
Do you also want to update `update_worker_pool_with_capacity_and_counts_by_state`, it still appears with sub-queries in the diff?

Reply:
I started to do so, but got `aggregate functions are not allowed in RETURNING`. I can continue to dig to see if I can figure it out, but also figured at the same time, this function will be way less used. What do you think?